### PR TITLE
fix(db-performance): improve sending and receiving msgs

### PR DESCRIPTION
**Pull request by mchenani**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?
During reviewing the performance of sending/receiving messages, we discovered some issues.
* **Inserting received messages:**
In this section, we found two major issues
* Fetching the userId and the ClientId
To send or insert a message, we always need the clientId [ to send a message] and the userId [for both processes received message and send a new message]. Fetching both these values was too expensive memory-wise! Since these two values are not changing after login, we decided to cache them during each user session and pass them to each usecase that needs them; these changes improved the process.
* Multiple queries to insert one message
To insert each received message, we were modifying the conversation to support the message unread counter and the notification flow. We put all of these queries in one DB transaction to do them in one DB connection.

### Issues

Improve performance.

### Causes (Optional)

Opening multiple DB connections and fetching user-scope static values multiple times.

### Solutions

Cache values and called related transactions in one single DB transaction.

### Dependencies (Optional)

_If there are some other pull requests related to this one (e.g. new releases of frameworks), specify them here._

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/message/MessageDAOImpl.kt [hunk 1]
```diff
@@ -33,8 +34,21 @@ class MessageDAOImpl(private val queries: MessagesQueries) : MessageDAO {
...
+            val selfUserId = QualifiedIDEntity("9371383d-a161-4428-a748-a87e8d0f00c7", "wire.com")
+            val isMessageComingFromOtherClient = message.senderUserId == selfUserId
+            if (isMessageComingFromOtherClient)
+                conversationsQueries.updateConversationReadDate(message.date, message.conversationId)
+            insertInDB(message)
+            if (shouldUpdateConversationOrder)
+                conversationsQueries.updateConversationModifiedDate(message.date, message.conversationId)
+            if (isMessageComingFromOtherClient)
+                conversationsQueries.updateConversationNotificationsDate(message.date, message.conversationId)
+        }
```
Comment (vitorhugods) [inline persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/message/MessageDAOImpl.kt:48]: I found out a cool feature of SQLDelight that might be better than just doing a transaction:

[SQLDelight group statements](https://cashapp.github.io/sqldelight/multiplatform_sqlite/grouping_statements/).

File: persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/message/MessageDAOImpl.kt [hunk 2]
```diff
@@ -33,8 +34,21 @@ class MessageDAOImpl(private val queries: MessagesQueries) : MessageDAO {
 
     override suspend fun deleteAllMessages() = queries.deleteAllMessages()
 
-    override suspend fun insertMessage(message: MessageEntity) = queries.transaction { insertInDB(message) }
+    override suspend fun insertMessage(message: MessageEntity, shouldUpdateConversationOrder: Boolean) {
+        queries.transaction {
+            val selfUserId = QualifiedIDEntity("9371383d-a161-4428-a748-a87e8d0f00c7", "wire.com")
```
Comment (vitorhugods) [inline persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/message/MessageDAOImpl.kt:39]: Eh... I think this one shouldn't be here 😅

File: persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/message/MessageDAOImpl.kt (hunk 1 above)
Comment (mchenani) [reply persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/message/MessageDAOImpl.kt:48]: Yes, we discussed it on Saturday with Tommaso and Mohammad, but it seems only works on one table at the same time!

Comment (mchenani) [reply persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/message/MessageDAOImpl.kt:48]: Just checked again, seems we have something like this:
```
updateSalaryRecord:
BEGIN;

UPDATE salary
SET salary=?
WHERE title=?;

INSERT INTO salaryChangeNotes(time, updateNote)
VALUES(?, ?);

END;
```
this is awesome! I'll try it.

File: persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/message/MessageDAOImpl.kt [hunk 3]
```diff
@@ -33,8 +34,20 @@ class MessageDAOImpl(private val queries: MessagesQueries) : MessageDAO {
 
     override suspend fun deleteAllMessages() = queries.deleteAllMessages()
 
-    override suspend fun insertMessage(message: MessageEntity) = queries.transaction { insertInDB(message) }
+    override suspend fun insertMessage(message: MessageEntity, selfUserId: UserIDEntity, shouldUpdateConversationOrder: Boolean) {
```
Comment (typfel) [inline persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/message/MessageDAOImpl.kt:37]: Thinking maybe we shouldn't move this business logic inside the persistence module but instead do something like this:

```
insertMessage(message: MessageEntity,  
              updateModifiedDate: Boolean = false, 
              updateReadDate: Boolean = false, 
              updateNotificationDate: Boolean: false) { ... }
```

And keep the logic in the logic module.

Comment (mchenani) [reply persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/message/MessageDAOImpl.kt:37]: Good point, will do that.

File: persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/message/MessageDAOImpl.kt [hunk 4]
```diff
@@ -33,8 +34,23 @@ class MessageDAOImpl(private val queries: MessagesQueries) : MessageDAO {
 
     override suspend fun deleteAllMessages() = queries.deleteAllMessages()
 
-    override suspend fun insertMessage(message: MessageEntity) = queries.transaction { insertInDB(message) }
+    override suspend fun insertMessage(
+        message: MessageEntity,
+        isMyMessage: Boolean,
```
Comment (typfel) [inline persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/message/MessageDAOImpl.kt:39]: This is also logic about inserting messages from the self user.  I would still suggest to break it up into a `updateConversationReadDate` and a `updateConversationNotificationDate`.

**Review by typfel — APPROVED** [review]

File: persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/message/MessageDAOImpl.kt [hunk 5]
```diff
@@ -33,8 +34,24 @@ class MessageDAOImpl(private val queries: MessagesQueries) : MessageDAO {
...
+        updateConversationNotificationsDate: Boolean
+    ) {
+        queries.transaction {
+            if (updateConversationReadDate)
+                conversationsQueries.updateConversationReadDate(message.date, message.conversationId)
+            insertInDB(message)
+            if (updateConversationModifiedDate)
+                conversationsQueries.updateConversationModifiedDate(message.date, message.conversationId)
+            if (updateConversationNotificationsDate)
+                conversationsQueries.updateConversationNotificationsDate(message.date, message.conversationId)
```
Comment (yamilmedina) [inline persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/message/MessageDAOImpl.kt:50]: Could we use some `{ }` here ? I think the `insertInDB(message)` may get lost while reading at first sight.

File: persistence/src/commonTest/kotlin/com/wire/kalium/persistence/dao/MessageDAOBenchmarkTest.kt [hunk 6]
```diff
@@ -67,6 +68,67 @@ class MessageDAOBenchmarkTest : BaseDatabaseTest() {
                         editStatus = MessageEntity.EditStatus.NotEdited
                     )
                 )
+
+                add(
```
Comment (yamilmedina) [inline persistence/src/commonTest/kotlin/com/wire/kalium/persistence/dao/MessageDAOBenchmarkTest.kt:72]: Nice to see this more cases added here 👌

**Review by yamilmedina — APPROVED** [review]
Small observation about readable part of if's. Great work @mchenani  👏

**Pull request merged**